From b76c2324cf46672db76c6f29f2331c29ba1f4b53 Mon Sep 17 00:00:00 2001
From: Devi Priya <quic_devipriy@quicinc.com>
Date: Sat, 14 Sep 2024 13:30:12 +0530
Subject: [PATCH] serial: qcom-geni: Skip loading GENI FW for console UART

The GENI FW for console UART would already be loaded by the
bootloaders. Hence Skip loading the FW for console UART as we see junk
characters when loading the FW while earlycon is enabled.

Change-Id: I4fe141dbec8d41f6c495b50331f5ed07d3ee818c
Signed-off-by: Devi Priya <quic_devipriy@quicinc.com>
---
 drivers/tty/serial/qcom_geni_serial.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/tty/serial/qcom_geni_serial.c b/drivers/tty/serial/qcom_geni_serial.c
index 05ab1bffe206..b735e6dcc4ae 100644
--- a/drivers/tty/serial/qcom_geni_serial.c
+++ b/drivers/tty/serial/qcom_geni_serial.c
@@ -189,14 +189,21 @@ static int qcom_geni_serial_request_port(struct uart_port *uport)
 {
 	struct platform_device *pdev = to_platform_device(uport->dev);
 	struct qcom_geni_serial_port *port = to_dev_port(uport);
+	const struct qcom_geni_device_data *data;
 
+	data = of_device_get_match_data(&pdev->dev);
+	if (!data)
+		return -EINVAL;
 	uport->membase = devm_platform_ioremap_resource(pdev, 0);
 	if (IS_ERR(uport->membase))
 		return PTR_ERR(uport->membase);
 	port->se.base = uport->membase;
 
 #ifdef CONFIG_QCOM_GENI_SE_FW_LOAD
-	geni_se_fw_load(&port->se, QUPV3_SE_UART);
+	if (!data->console)
+		geni_se_fw_load(&port->se, QUPV3_SE_UART);
+	else
+		pr_info("Skipping GENI FW load for console UART\n");
 #endif /* CONFIG_QCOM_GENI_SE_FW_LOAD */
 
 	return 0;
-- 
2.34.1

